README.md
# Signum Node
[![Node Build](https://github.com/signum-network/signum-node/actions/workflows/build.yml/badge.svg)](https://github.com/signum-network/signum-node/actions/workflows/build.yml)
[![GPLv3](https://img.shields.io/badge/license-GPLv3-blue.svg)](LICENSE.txt)
[![Get Support at https://discord.gg/ms6eagX](https://img.shields.io/badge/join-discord-blue.svg)](https://discord.gg/ms6eagX)

The world's first HDD-mined cryptocurrency using an energy efficient
and fair Proof-of-Commitment (PoC+) consensus algorithm running since August 2014.

The two supported database backends are:

- H2 (embedded, recommended)
- MariaDB (advanced users)

## Network Features

- Proof of Commitment - ASIC proof / Energy efficient and sustainable mining
- No ICO/Airdrops/Premine
- Turing-complete smart contracts, via [Signum SmartJ](https://github.com/signum-network/signum-smartj)
- Asset Exchange; Digital Goods Store; Crowdfunds, NFTs, games, and more (via smart contracts); and Alias system

## Network Specification

- Block time is 4 minutes 
- Block size is 375,360 byte
- Minimum transaction size is 184 bytes
- Minimum network fee is 0.01 Signa
- TPS is 16
- STPS is up to 5,000 [STPS](https://docs.signum.network/signum#ny-smart-layer-10)
- Maximum 1,200,000 balance changes per block
- Total Supply: [2,138,119,200 SIGNA up to block 972k + 100 SIGNA per block after that](https://github.com/signum-network/CIPs/blob/master/cip-0029.md)
- Block reward started at 10,000/block in 2014
- Block reward decreased at 5% each month with a minimum mining incentive of 100 SIGNA per block
- Automated burning of subscription interval payment fees and smart contract step fees from block 1,029,000 [SIP-36](https://github.com/signum-network/SIPs/blob/master/SIP/sip-36.md)

## Features

- Decentralized Peer-to-Peer network with spam protection
- Built in Java - runs anywhere, from a Raspberry Pi to a Phone
- Fast sync with multithreaded CPU or, optionally, an OpenCL GPU
- HTTP API for clients to interact with network

# Installation

## Prerequisites

### Windows

Any recent 64 bit Windows should suffice (a Java 11 is embedded in the windows package).

### Linux and Mac, Java 64-bit 11 (Recommended) or higher

You need Java 64-bit 11 (recommended) or higher installed.
Install the `openjdk-11-jre` package or similar for your distribution.
To check your java version, run `java -version`. You should get an output similar to the following:

```text
openjdk version "11.0.13" 2021-10-19
OpenJDK Runtime Environment (build 11.0.13+8-Ubuntu-0ubuntu1.20.04)
OpenJDK 64-Bit Server VM (build 11.0.13+8-Ubuntu-0ubuntu1.20.04, mixed mode, sharing)
```

The important part is that the Java version starts with `11.` (Java 11)

### MariaDB (Optional)

[Download and install MariaDB](https://mariadb.com/downloads/mariadb-tx)

The MariaDb installation will ask to setup a password for the root user. 
Add this password to the `node.properties` file you will create when installing BRS:

```properties
DB.Url=jdbc:mariadb://localhost:3306/signum
DB.Username=root
DB.Password=YOUR_PASSWORD
```

## Set Up

Grab the latest [release](https://github.com/signum-network/signum-node/releases) (or, if you prefer, compile yourself using the instructions below)

In the `conf` directory, copy `node-default.properties` into a new file named `node.properties` and modify this file to suit your needs (See "Configuration" section below)

To run the node, double click on `signum-node.exe` (if on Windows) or run `java -jar signum-node.jar`.
On most systems this will show you a monitoring window and will create a tray icon to show that Signum node is running. To disable this, instead run `java -jar signum-node.jar --headless`.

## Configuration

### Running on mainnet

Starting with the release version 3.3.0 and higher the concept of the config file changed.
To run the node for mainnet with the default options, no configuration change is needed.

All default/recommended parameters are defined in code by the Signum Node but you can overwrite the parameter within the config file to suit your needs.
The default values for the available settings are shown in the `conf/node-default.properties` file as commented out. 

### Configuration Hints

**Configure your cash-back**

As an incentive for users to run their own full nodes, [SIP-35](https://github.com/signum-network/SIPs/blob/master/SIP/sip-35.md) introduced the concept of fee *cash-back*.
In order to receive the 25% cashback on the fees for transactions created by your node, set your
own account ID in the configuration file:

```properties
node.cashBackId = 8952122635653861124
```
Note: the example ID above `8952122635653861124` is the [SNA](https://www.sna.signum.network/) account.

The cash-back is paid from block 1,029,000.

**H2/MariaDB**

By default Signum Node is using H2 (file based) as database. 
If you like to use MariaDB you will need to adjust your `conf/node.properties`:

```properties
#### DATABASE ####
DB.Url=jdbc:mariadb://localhost:3306/signum
DB.Username=
DB.Password=
```

Please modify the `DB.Url` to your own specifications (port 3306 is the standard port from MariaDB) and also set the `DB.Username` and `DB.Password` according your setup for the created database.

**UPnP-Portforwarding**

By default the UPnP port forwarding is activated. 
If you run the node on a VPS you can deactivate this by setting it to "no".
```properties
## Port for incoming peer to peer networking requests, if enabled.
# P2P.Port = 8123
## Use UPnP-Portforwarding
P2P.UPnP = no
```

**Enable SNR**

If you set on the `P2P.myPlatform` a valid Signum address and you fulfill the SNR requirements your node aka the set Signum address will be rewarded with the SNR.
The SNR ([Signum Network Reward](https://signum.community/signum-snr-awards/)) is a community driven bounty paid to all node-operators which run continuous (uptime > 80%) and with the newest release a Signum node.
```properties
## My platform, to be announced to peers.
## Enter your Signum address here for SNR rewards, see here: https://signum.community/signum-snr-awards/
P2P.myPlatform = S-ABCD-EFGH-IJKL-MNOP
```
You can check your node using the [explorer](https://explorer.signum.network/peers/).

## Hardware Requirements

The Signum Node is not hardware demanding nor it needs a fast internet connection.
The specifications for the hardware is as follows:

-   Minimum : 1 vCPU, 1 GB RAM, 20 GB HD,  Minimum Swapfile 2GB (Linux)
-   Recommended : 2 vCPU, 2 GB RAM, 20 GB HD, Minimum Swapfile 4GB (Linux)

**Tuning options**

If you run the minimum requirement you can turn off the `indirectIncomingService` in the config file to reduce CPU and file usage. By default this parameter is activated.
```properties
## Enable the indirect incoming tracker service. This allows you to see transactions where you are paid
## but are not the direct recipient eg. Multi-Outs.
node.indirectIncomingService.enable = false
```


## Testnet

Starting with the Signum node version 3.3.0 and higher the node is able to handle different chains in a multiverse manner. All parameters for a different chain are set in the code section of the node and can be activated by pointing to the other chain aka universe in the config file.
No additional setting is needed. 

In order to run a testnet node adjust your `conf/node.properties` to:
```properties
## Run with a different network
# Testnet network
node.network = signum.net.TestnetNetwork
```

If no custom DB is set, a H2 file will be created as `db/signum-testnet.mv.db`.
For a MariaDB setup you need to configure a testnet instance in the config file.

### Private Chains

In order to run a private (local) chain with *mock* mining just select the network:

```properties
node.network = signum.net.MockNetwork
```

This will allow you to forge new blocks as soon as you submit a new nonce.
Note that P2P is disabled when running in this mode.

# API Documentation

Since Version 3.4.3 the new interactive API documentation based on [OpenAPI Spec V3](https://www.openapis.org/) is shipped and active by default.

You can access it via `<host>:8125/api-doc` (or `<host>:6876/api-doc` for Testnet respectively).

[Read more about the documentation](/openapi/README.md) 
 
# Building from sources

## Building the latest stable release

Run these commands (the `main` branch is always the latest stable release):

```bash
git clone https://github.com/signum-network/signum-node.git
cd signum-node
./gradlew dist
```

Your packaged release will now be available in the `build/distribution` directory.

## Building the latest development version

Clone the repository as instructed above and run these commands:

```bash
git switch develop
./gradlew dist
```

Your packaged release will now be available in the `build/distribution` directory.

**Please note that development builds will refuse to run outside of testnet or a private chain**

## Running the automated tests

Clone the repository as instructed above and run:

```bash
./gradlew test
```

## Updating the Phoenix Wallet

Since V3.0 the Phoenix Wallet is available as built-in alternative to the classic wallet. 
Within a release of the node software automatically the latest available release of the Phoenix wallet will be applied.
As the Phoenix Wallet is a project apart from this repository the node and wallet software have different
release cycles. Therefore, an additional update script (at this moment only for Linux/MacOS) is provided.

Just run `./update-phoenix.sh`, which is available in the distribution package 

# Releasing

To cut a new (pre)-release just create a tag of the following format `vD.D.D[-suffix]`. Githubs actions automatically creates
a pre-release with entirely build executable as zip.

```bash
git tag v3.0.1-beta
git push --tags
```

# Docker

See DOCKER.md for information on running and building docker images.

# Developers

Main Developer: [jjos2372](https://github.com/jjos2372). Donation address: [S-JJQS-MMA4-GHB4-4ZNZU](https://explorer.signum.network/?action=account&account=3278233074628313816)
Frequent Contributors: 
- [ohager](https://github.com/ohager). Donation address: [S-9K9L-4CB5-88Y5-F5G4Z](https://explorer.signum.network/?action=account&account=16107620026796983538)
- [damccull](https://github.com/damccull).

For more information, see [Credits](doc/Credits.md)

# Appendix

* [Version History](doc/History.md)

* [Credits](doc/Credits.md)

* [References/Links](doc/References.md)


doc/History.md
[History of Signum](https://signum.community/signum-development/)

```
2020-04-06 v3.0.0
- Proof of Commitment (PoC+) implemented - CIP27
- Support for extended addresses (no more *activation* for new addresses) - CIP26
- Mininum block reward limit - CIP29
- Increase max steps for smart contracts (ATs) - CIP28
- Phoenix wallet included
- Classic wallet updated
- Suggested fees based on unconfirmed transactions and not transaction history
- Improved GUI with new theme and buttons
- API documentation under `/api-doc` and not `/test`
- Difficulty given by network capacity + amount committed
- Fixed bugs with the donwload cache
- Fixed bugs preventing to reconnect after network outage
- Fix bug with AT on mining nodes (making them occasionaly inconsistent)
- New transaction types for commitment add/remove
- Improved popoff functions
- Faster and better DB consistency checks
- Better DB performance with new indices added to some tables
- Peers below v3.0 automatically disconnected after Signum fork block
- Several debug messages added for diagnose possible future problems
- Assets renamed to Tokens
- Unused features were removed from the classic wallet (still available by API)
- Peers now share their base targets so any inconsistency will be detected immediately
- Fixed pagination for multi-out transactions
- More AT bugs fixed with the commands SET_IDX and IDX_DAT
- A working *reftx* implementation
- Several libraries were updated (including jetty and other DB related)
- More stats available on the `getState` API

2020-01-07 v2.5.4
- Faster and more precise database check based on the balance of all accounts, total burnt, and total BURST ever mined
- Extended API to register bigger ATs (smart contracts)
- New buttons to pop-off blocks on the GUI (when the debug flag is enabled)

2020-10-06 v2.5.3
- Nodes that experienced an unclean kill, power outage, or crash can have the database in an inconsistent state. This version contains an additional database check based on the balance of all accounts and total BURST ever mined.

2020-08-29 v2.5.2
- Fixes an issue which caused excessive CPU load when validating transactions with very high fees

2020-05-31 v2.5.1
- Docker build
- Windows executable now refuse to run on 32 bit Java and a download link is shown to the user
- Limit the amount of blocks sent to peers by total length and not number of blocks
- Improved command line arguments, including one for setting the conf folder
- Other minor improvements

2020-05-10 v2.5.0 (Backport of fixes from v3.0.0 and many other improvements)
- More powerful and cheaper running smart contracts (CIP20)
- Deeplink Generator API (CIP22)
- Enforce slot fees (CIP23)
- New deadline computation formula aiming at more stable block times (CIP24)
- Can run on recent Java versions (previously limited to Java 8)
- A new GUI implementation
- Improved CORS support
- Fixed Escrow ID in Escrow Result Attachment Protobuf is always 2
- Fixed Get Peer Handler in gRPC API is never initialized
- Fixed GetOrdersHandler always returns ask orders
- Fixed Bugs in AT API Implementation

2019-08-13 v2.4.2
- Fixed HTTP API Encoding being reported to the client incorrectly, leading to the client incorrectly parsing special characters
- Limit maximum number of items returned by HTTP API
- Fixed NPE in ProtoBuilder when fetching an account

2019-07-15 v2.4.1
- Default to submit nonce whitelist off
- Revert removal of rejection of surplus parameters
- Add option to bind V2 API to specific interface
- MariaDB Settings tweaks 
- Various bug fixes and improvements

2019-07-01 v2.4.0
- Massive DB optimization, much much faster sync speed (Benchmarked at 7 hours to sync to block 600k on a 4C/8T 16GB RAM system, under MariaDB 10.3)
- Implemented CIP19 - View incoming & outgoing multi-out transactions in the UI
- Added new feature to sign arbitrary messages using UI
- Fixed gRPC error descriptions
- Comprehensive V2 API with all functionality of V1 implemented
- Auto pop-off on block push fail with slow back-off, should prevent nodes from getting stuck forever
- UTStore should produce waaaay less spam
- CORS on by default
- Minimum previous version is now v2.3.0
- Enforce fee structure (Inactive)
- Improved algorithm for transaction candidate selection
- Check in gRPC generated files (simplifies build)
- Tighter timings for sync threads
- Burstkit4j integration
- Rewrite support for UI (Apps that utilize deep linking such as phoenix can now be hosted by BRS)
- Add a method to not submit passphrase when solo mining by configuring passphrase in config and only submitting account ID, and an option to disallow others from mining on your node
- AT debug option
- Improvements to AT implementation
- Web UI: Display AT messages as both string and hex
- Fix UT Store failed removal
- Re-add `getGuaranteedBalance` HTTP API call as lots of clients depended on it
- Test endpoint support for QR code generator
- Implemented CIP20 (Inactive)

2018-04-04 v2.3.0
           Fix of major security vulnerability where passphrase was sent to node upon login
           gRPC-based V2 API. Currently only contains calls needed for mining, will be expanded in future if well received.
           Migrate to GSON as JSON library
           Significantly improve sync speed, as well as other minor performance improvements
           New Semver-based versioning system
           Fix bug where reward recipient assignments would not go into unconfirmed transactions
           Lightweight Desktop GUI, with tray icon (For windows and mac, can be disabled with "--headless" command line argument)
           Automatically add conf/ directory to classpath
           Configurable TestNet UI/API port
           New getAccountsWithName API call
           UI: Fix 24h timestamp display option
           Allow development versions of wallet to run on TestNet only
           Fixed bug where string validation could fail in certain locales
           Use FlywayDB for database migration management

src/brs/Block.java
package brs;

import brs.crypto.Crypto;
import brs.db.TransactionDb;
import brs.fluxcapacitor.FluxValues;
import brs.peer.Peer;
import brs.util.Convert;
import brs.util.JSON;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.*;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;

public class Block {

  private static final Logger logger = LoggerFactory.getLogger(Block.class);
  private final int version;
  private final int timestamp;
  private final long previousBlockId;
  private final byte[] generatorPublicKey;
  private final byte[] previousBlockHash;
  private final long totalAmountNQT;
  private final long totalFeeNQT;
  private final long totalFeeBurntNQT;
  private final long totalFeeCashBackNQT;
  private final int payloadLength;
  private final byte[] generationSignature;
  private final byte[] payloadHash;
  private final AtomicReference<List<Transaction>> blockTransactions = new AtomicReference<>();

  private byte[] blockSignature;

  private BigInteger cumulativeDifficulty = BigInteger.ZERO;

  private long baseTarget = Constants.INITIAL_BASE_TARGET;
  private final AtomicLong nextBlockId = new AtomicLong();
  private int height = -1;
  private final AtomicLong id = new AtomicLong();
  private final AtomicReference<String> stringId = new AtomicReference<>();
  private final AtomicLong generatorId = new AtomicLong();
  private long nonce;

  private BigInteger pocTime = null;
  private long commitment = 0L;

  private final byte[] blockATs;

  private Peer downloadedFrom = null;
  private int byteLength = 0;

  Block(int version, int timestamp, long previousBlockId, long totalAmountNQT, long totalFeeNQT,
      long totalFeeCashBackNQT, long totalFeeBurntNQT,
      int payloadLength, byte[] payloadHash, byte[] generatorPublicKey, byte[] generationSignature,
      byte[] blockSignature, byte[] previousBlockHash, List<Transaction> transactions,
      long nonce, byte[] blockATs, int height, long baseTarget) throws BurstException.ValidationException {

    if (payloadLength > Burst.getFluxCapacitor().getValue(FluxValues.MAX_PAYLOAD_LENGTH, height) || payloadLength < 0) {
      throw new BurstException.NotValidException(
          "attempted to create a block with payloadLength " + payloadLength + " height " + height + "previd " + previousBlockId);
    }

    this.version = version;
    this.timestamp = timestamp;
    this.previousBlockId = previousBlockId;
    this.totalAmountNQT = totalAmountNQT;
    this.totalFeeNQT = totalFeeNQT;
    this.totalFeeCashBackNQT = totalFeeCashBackNQT;
    this.totalFeeBurntNQT = totalFeeBurntNQT;
    this.payloadLength = payloadLength;
    this.payloadHash = payloadHash;
    this.generatorPublicKey = generatorPublicKey;
    this.generationSignature = generationSignature;
    this.blockSignature = blockSignature;

    this.previousBlockHash = previousBlockHash;
    if (transactions != null) {
      this.blockTransactions.set(Collections.unmodifiableList(transactions));
      if (blockTransactions.get().size() > (Burst.getFluxCapacitor().getValue(FluxValues.MAX_NUMBER_TRANSACTIONS, height))) {
        throw new BurstException.NotValidException(
            "attempted to create a block with " + blockTransactions.get().size() + " transactions");
      }
      long previousId = 0;
      for (Transaction transaction : this.blockTransactions.get()) {
        if (transaction.getId() <= previousId && previousId != 0) {
          throw new BurstException.NotValidException("Block transactions are not sorted!");
        }
        previousId = transaction.getId();
      }
    }
    this.nonce = nonce;
    this.blockATs = blockATs;
    this.baseTarget = baseTarget;
  }

  public Block(int version, int timestamp, long previousBlockId, long totalAmountNQT, long totalFeeNQT,
      long totalFeeCashBackNQT, long totalFeeBurntNQT,
      int payloadLength, byte[] payloadHash, byte[] generatorPublicKey, byte[] generationSignature, byte[] blockSignature, byte[] previousBlockHash, BigInteger cumulativeDifficulty, long baseTarget,
      long nextBlockId, int height, Long id, long nonce, byte[] blockATs) throws BurstException.ValidationException {

    this(version, timestamp, previousBlockId, totalAmountNQT, totalFeeNQT, totalFeeCashBackNQT, totalFeeBurntNQT, payloadLength, payloadHash, generatorPublicKey, generationSignature, blockSignature, previousBlockHash, null, nonce, blockATs, height, baseTarget);

    this.cumulativeDifficulty = cumulativeDifficulty == null ? BigInteger.ZERO : cumulativeDifficulty;
    this.nextBlockId.set(nextBlockId);
    this.height = height;
    this.id.set(id);
  }

  private TransactionDb transactionDb() {
    return Burst.getDbs().getTransactionDb();
  }

  public boolean isVerified() {
    return pocTime != null;
  }

  public void setPeer(Peer peer) {
    this.downloadedFrom = peer;
  }

  public Peer getPeer() {
    return this.downloadedFrom;
  }

  public void setByteLength(int length) {
    this.byteLength = length;
  }

  public int getByteLength() {
    return this.byteLength;
  }

  public int getVersion() {
    return version;
  }

  public int getTimestamp() {
    return timestamp;
  }

  public long getPreviousBlockId() {
    return previousBlockId;
  }

  public byte[] getGeneratorPublicKey() {
    return generatorPublicKey;
  }

  public byte[] getBlockHash() {
    return Crypto.sha256().digest(getBytes());
  }

  public byte[] getPreviousBlockHash() {
    return previousBlockHash;
  }

  public long getTotalAmountNQT() {
    return totalAmountNQT;
  }

  public long getTotalFeeNQT() {
    return totalFeeNQT;
  }

  public long getTotalFeeCashBackNQT() {
    return totalFeeCashBackNQT;
  }

  public long getTotalFeeBurntNQT() {
    return totalFeeBurntNQT;
  }

  public int getPayloadLength() {
    return payloadLength;
  }

  public byte[] getPayloadHash() {
    return payloadHash;
  }

  public byte[] getGenerationSignature() {
    return generationSignature;
  }

  public byte[] getBlockSignature() {
    return blockSignature;
  }

  public List<Transaction> getTransactions() {
    if (blockTransactions.get() == null) {
      this.blockTransactions.set(Collections.unmodifiableList(transactionDb().findBlockTransactions(getId(), true)));
      this.blockTransactions.get().forEach(transaction -> transaction.setBlock(this));
    }
    return blockTransactions.get();
  }

  public List<Transaction> getAllTransactions() {
    return Collections.unmodifiableList(transactionDb().findBlockTransactions(getId(), false));
  }

  public long getBaseTarget() {
    return baseTarget;
  }

  public long getCapacityBaseTarget() {
    long capacityBaseTarget = baseTarget;
    if(Burst.getFluxCapacitor().getValue(FluxValues.POC_PLUS, height)) {
      // Base target encoded as two floats, one for the commitment and the other the classical base target
      float capacityBaseTargetFloat = Float.intBitsToFloat((int)(baseTarget & 0xFFFFFFFFL));
      capacityBaseTarget = (long)capacityBaseTargetFloat;
    }
    return capacityBaseTarget;
  }

  public long getAverageCommitment() {
    if(Burst.getFluxCapacitor().getValue(FluxValues.POC_PLUS, height)) {
      // Base target encoded as two floats, one for the commitment and the other the classical base target
      float commitmentBaseTargetFloat = Float.intBitsToFloat((int)((baseTarget) >> 32));
      return (long)commitmentBaseTargetFloat;
    }
    return Constants.INITIAL_COMMITMENT;
  }

  public void setBaseTarget(long baseTargetCapacity, long averageCommitment) {
    this.baseTarget = ((long)Float.floatToIntBits((float)averageCommitment)) << 32 |
        ((long)Float.floatToIntBits((float)baseTargetCapacity));
  }

  public BigInteger getCumulativeDifficulty() {
    return cumulativeDifficulty;
  }

  public long getNextBlockId() {
    return nextBlockId.get();
  }

  public int getHeight() {
    return height;
  }

  public void setHeight(int height) {
    this.height = height;
  }

  public long getId() {
    if (id.get() == 0) {
      if (blockSignature == null) {
        throw new IllegalStateException("Block is not signed yet");
      }
      byte[] hash = Crypto.sha256().digest(getBytes());
      long longId = Convert.fullHashToId(hash);
      id.set(longId);
      stringId.set(Convert.toUnsignedLong(longId));
    }
    return id.get();
  }

  public String getStringId() {
    if (stringId.get() == null) {
      getId();
      if (stringId.get() == null) {
        stringId.set(Convert.toUnsignedLong(id.get()));
      }
    }
    return stringId.get();
  }

  public long getGeneratorId() {
    if (generatorId.get() == 0) {
      generatorId.set(Account.getId(generatorPublicKey));
    }
    return generatorId.get();
  }

  public Long getNonce() {
    return nonce;
  }

  public boolean equals(Object o) {
    return o instanceof Block && this.getId() == ((Block) o).getId();
  }

  public int hashCode() {
    return (int) (getId() ^ (getId() >>> 32));
  }

  public JsonObject getJsonObject() {
    JsonObject json = new JsonObject();
    json.addProperty("version", version);
    json.addProperty("timestamp", timestamp);
    json.addProperty("previousBlock", Convert.toUnsignedLong(previousBlockId));
    json.addProperty("totalAmountNQT", totalAmountNQT);
    json.addProperty("totalFeeNQT", totalFeeNQT);
    json.addProperty("totalFeeCashBackNQT", totalFeeCashBackNQT);
    json.addProperty("totalFeeBurntNQT", totalFeeBurntNQT);
    json.addProperty("payloadLength", payloadLength);
    json.addProperty("payloadHash", Convert.toHexString(payloadHash));
    json.addProperty("generatorPublicKey", Convert.toHexString(generatorPublicKey));
    json.addProperty("generationSignature", Convert.toHexString(generationSignature));
    if (version > 1) {
      json.addProperty("previousBlockHash", Convert.toHexString(previousBlockHash));
    }
    json.addProperty("blockSignature", Convert.toHexString(blockSignature));
    JsonArray transactionsData = new JsonArray();
    getTransactions().forEach(transaction -> transactionsData.add(transaction.getJsonObject()));
    json.add("transactions", transactionsData);
    json.addProperty("nonce", Convert.toUnsignedLong(nonce));
    json.addProperty("baseTarget", Convert.toUnsignedLong(baseTarget));
    json.addProperty("blockATs", Convert.toHexString(blockATs));
    return json;
  }

  static Block parseBlock(JsonObject blockData, int height) throws BurstException.ValidationException {
    try {
      int version = JSON.getAsInt(blockData.get("version"));
      int timestamp = JSON.getAsInt(blockData.get("timestamp"));
      long previousBlock = Convert.parseUnsignedLong(JSON.getAsString(blockData.get("previousBlock")));
      long totalAmountNQT = JSON.getAsLong(blockData.get("totalAmountNQT"));
      long totalFeeNQT = JSON.getAsLong(blockData.get("totalFeeNQT"));
      long totalFeeCashBackNQT = 0L;
      long totalFeeBurntNQT = 0L;
      if (version > 3){
        totalFeeCashBackNQT = JSON.getAsLong(blockData.get("totalFeeCashBackNQT"));
        totalFeeBurntNQT = JSON.getAsLong(blockData.get("totalFeeBurntNQT"));
      }
      int payloadLength = JSON.getAsInt(blockData.get("payloadLength"));
      byte[] payloadHash = Convert.parseHexString(JSON.getAsString(blockData.get("payloadHash")));
      byte[] generatorPublicKey = Convert.parseHexString(JSON.getAsString(blockData.get("generatorPublicKey")));
      byte[] generationSignature = Convert.parseHexString(JSON.getAsString(blockData.get("generationSignature")));
      byte[] blockSignature = Convert.parseHexString(JSON.getAsString(blockData.get("blockSignature")));
      byte[] previousBlockHash = version == 1 ? null : Convert.parseHexString(JSON.getAsString(blockData.get("previousBlockHash")));
      long nonce = Convert.parseUnsignedLong(JSON.getAsString(blockData.get("nonce")));
      long baseTarget = Convert.parseUnsignedLong(JSON.getAsString(blockData.get("baseTarget")));

      if(Burst.getFluxCapacitor().getValue(FluxValues.POC_PLUS, height) && baseTarget == 0L) {
        throw new BurstException.NotValidException("Block received without a baseTarget");
      }

      SortedMap<Long, Transaction> blockTransactions = new TreeMap<>();
      JsonArray transactionsData = JSON.getAsJsonArray(blockData.get("transactions"));

      for (JsonElement transactionData : transactionsData) {
        Transaction transaction = Transaction.parseTransaction(JSON.getAsJsonObject(transactionData), height);
        if (transaction.getSignature() != null && blockTransactions.put(transaction.getId(), transaction) != null) {
          throw new BurstException.NotValidException("Block contains duplicate transactions: " + transaction.getStringId());
        }
      }

      byte[] blockATs = Convert.parseHexString(JSON.getAsString(blockData.get("blockATs")));
      return new Block(version, timestamp, previousBlock, totalAmountNQT, totalFeeNQT,
          totalFeeCashBackNQT, totalFeeBurntNQT,
          payloadLength, payloadHash, generatorPublicKey, generationSignature, blockSignature,
          previousBlockHash, new ArrayList<>(blockTransactions.values()), nonce, blockATs, height, baseTarget);
    } catch (BurstException.ValidationException | RuntimeException e) {
      if (logger.isDebugEnabled()) {
        logger.debug("Failed to parse block: {}", JSON.toJsonString(blockData));
      }
      throw e;
    }
  }

  public byte[] getBytes() {
    ByteBuffer buffer = ByteBuffer.allocate(4 + 4 + 8 + 4 + (version < 3 ? (4 + 4) : (8 + 8)) + 4
        + 32 + 32 + (32 + 32) + 8 + (blockATs != null ? blockATs.length : 0) + 64);
    buffer.order(ByteOrder.LITTLE_ENDIAN);
    buffer.putInt(version);
    buffer.putInt(timestamp);
    buffer.putLong(previousBlockId);
    buffer.putInt(getTransactions().size());
    if (version < 3) {
      buffer.putInt((int) (totalAmountNQT / Constants.ONE_BURST));
      buffer.putInt((int) (totalFeeNQT / Constants.ONE_BURST));
    } else {
      buffer.putLong(totalAmountNQT);
      buffer.putLong(totalFeeNQT);
    }
    buffer.putInt(payloadLength);
    buffer.put(payloadHash);
    buffer.put(generatorPublicKey);
    buffer.put(generationSignature);
    if (version > 1) {
      buffer.put(previousBlockHash);
    }
    buffer.putLong(nonce);
    if (blockATs != null)
      buffer.put(blockATs);
    if (buffer.limit() - buffer.position() < blockSignature.length)
      logger.error("Something is too large here - buffer should have {} bytes left but only has {}",
                   blockSignature.length,
                   (buffer.limit() - buffer.position()));
    buffer.put(blockSignature);
    return buffer.array();
  }

  void sign(String secretPhrase) {
    if (blockSignature != null) {
      throw new IllegalStateException("Block already signed");
    }
    blockSignature = new byte[64];
    byte[] data = getBytes();
    byte[] data2 = new byte[data.length - 64];
    System.arraycopy(data, 0, data2, 0, data2.length);
    blockSignature = Crypto.sign(data2, secretPhrase);
  }

  public byte[] getBlockATs() {
    return blockATs;
  }

  public BigInteger getPocTime() {
    return pocTime;
  }

  public void setPocTime(BigInteger pocTime) {
    this.pocTime = pocTime;
  }

  public long getCommitment() {
    return this.commitment;
  }

  public void setCommitment(long commitment) {
    this.commitment = commitment;
  }

  public void setBaseTarget(long baseTarget) {
    this.baseTarget = baseTarget;
  }

  public void setCumulativeDifficulty(BigInteger cumulativeDifficulty) {
    this.cumulativeDifficulty = cumulativeDifficulty;
  }
}
